[PREVIEW] Remove mat views, use counter tables, and fix doubling counting

kevinhartman opened the pull request:
Not ready to merge.

# Changes
- Materialized Views have been replaced with simple tables in an effort to:
  - support portability to other Cassandra API consistent DBs such as Cosmos.
  - allow counter columns for aggregations to improve data consistency.
- Aggregation tables have been made to use the same Partition Keys in an effort to colocate Spark partition data with corresponding Cassandra partitions. `TileRow` data (the base unit of most aggregations) is partitioned once in Spark, colocating it with destination Cassandra partitions. As a result, writing table data should avoid write shuffling for most aggregations. Note that some aggregation tables were sufficiently different such that they could not reasonably be made to use the same partition key.
- Event de-duplication previously handled by the event batch materialized view has been replaced with a **left excluding join** from the current Spark batch to the `events` table ([reference](https://www.codeproject.com/Articles/33052/Visual-Representation-of-SQL-Joins)).
- Aggregation row generators for `CassandraPopularPlaces `, `CassandraTileBucket `, `CassandraHeatmapTiles `, and `CassandraEventPlacesSchema ` have been consolidated into a single generator, `TileRows`. With the old generators, the same rows were being computed multiple times, but with slightly different fields. `TileRows` generates a super set of all rows previously generated, and includes all fields needed by downstream tables on each row. Unneeded information in a row is eliminated automatically when saving to target tables which don't include corresponding columns. In some cases, some rows are not applicable to a target table (i.e. `populartopics` which must not contain rows with > 1 conjunctive topic), in which case these rows are filtered before saving using Spark.
- A double-counting issue for aggregations (mention count as well as avg sentiment numerator) has been resolved through the introduction of `CassandraExtensions.deDupValuesByCassandraTable`, which consolidates the generated tile rows for a batch which would unintentionally collide given the target table's primary key, prior to saving. For example, previously, the generated `ComputedTile` rows from an event were unique when written to the `heatmap` table, but would collide when written to `computedtiles` since its table's primary key is less specific. Counter table collisions lead to double counting. The general deduplication approach for counter tables taken here is to increment any given aggregation table row at most once per event (a single event can increment multiple rows, and a single row can be incremented by multiple distinct events, but a row cannot be incremented multiple times by a single event).

# Outstanding
- Since mention count columns are now counter columns, query results can no longer be ordered by mention count. This affects `populartopics` and `popularsources`, which use `mentioncount` as a *secondary* ordering to `perioddate`. If mention count is important to the Fortis interface, code will need to be added to order by mention count within the client.
- Test code needs to be written.
- A few TODO questions throughout PR.

kevinhartman committed:
Remove materialized views and use counters for all aggregations.

kevinhartman committed:
Add note on cleaning to docker docs.

kevinhartman committed:
[DROP] set replication to 1 for tests.

kevinhartman committed:
Cache event dedup result to fix bug where they'd be recalculated after writing to events table.

kevinhartman committed:
Make TileAggregator interface more testable and move table names to constants.

kevinhartman committed:
Add deduplication when saving to Cassandra.

kevinhartman committed:
Perform deduplication only within the scope of a single event.

kevinhartman committed:
Reorder cassandra setup CQL to improve diff.

c-w commented on project-fortis-pipeline/docs/development-setup.md:
```diff
@@ -113,6 +113,14 @@ Note that any changes to the React code in project-fortis-interfaces folder will
 be automatically detected and re-loaded so the re-build step above won't be
 necessary for changes to the frontend.
 
+## Cleaning
```
The developer FAQ has moved. To prevent re-work due to merges, I've updated the new location with this snippet, so you can remove this from your PR: https://github.com/CatalystCode/project-fortis/commit/0f2d35181104548e1e46ff0f0d1c0164623de2c5

c-w replied:
Done in https://github.com/CatalystCode/project-fortis/pull/31/commits/80c7ffa81afaad246c64a19ac40bd012d44de266

c-w commented on project-fortis-pipeline/ops/storage-ddls/cassandra-setup.cql:
```diff
@@ -1,75 +1,82 @@
 CREATE KEYSPACE fortis WITH replication = {
     'class': 'SimpleStrategy',
-    'replication_factor': 3
+    'replication_factor': 1 // TODO: change to 3 before merging
```
Reminder to not check this in.

Note that the run-node.sh file should already be adapting the replication factor to 1 since the appropriate environment variable is being set in docker-compose.yml

https://github.com/CatalystCode/project-fortis/blob/0f2d35181104548e1e46ff0f0d1c0164623de2c5/docker-compose.yml#L42

https://github.com/CatalystCode/project-fortis/blob/0f2d35181104548e1e46ff0f0d1c0164623de2c5/project-fortis-services/docker/run-node.sh#L30-L35

c-w commented on project-fortis-pipeline/ops/storage-ddls/cassandra-setup.cql:
```diff
@@ -120,83 +127,69 @@ CREATE TABLE fortis.events(
+    conjunctiontopic3 text,
+    tilez int,
+    tileid text,
+    externalsourceid text, // TODO: should this be part of the Primary Key?
```
This table is used to drive the facts page which displays a list of all the events that happened, broken down by pipeline and filterable by keyword and date. As such, I don't think the externalsourceid is required as part of the key: we essentially just need a way to get the eventids for a given set of keywords, pipeline and date.

kevinhartman replied:
Makes sense. My concern was that it'd be possible for rows with the same primary key to collide even if they had different `externalsourceid`s, but I suppose we'd never see a case where two inserts could have the same `eventid` but different `externalsourceid`.

c-w commented on project-fortis-spark/src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventSchema.scala:
```diff
@@ -70,99 +61,65 @@ object CassandraConjunctiveTopics {
-      placeid = location.placeid
-    )
+        avgsentimentnumerator = (computedfeatures.sentiment.neg_avg * DoubleToLongConversionFactor).toLong,
+        insertiontime = new Date().getTime
```
This might be more explicit in the fact that we're using UTC time: `Instant.now().toEpochMillis()`
Alternatively, plain old `System.getCurrentTimeMillis()` should also do the trick.

c-w commented on project-fortis-spark/src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala:
```diff
@@ -83,43 +80,28 @@ object CassandraEventsSink {
         consistencyLevel = ConsistencyLevel.ALL // Ensure write is consistent across all replicas
```
Would local quorum consistency also be good enough here? We're now using local quorum for reads where previously we were using local one, so we should be able to get consistent read behavior with lower write consistency.

kevinhartman replied:
Local quorum it is!

c-w commented on project-fortis-spark/src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/CassandraEventsSink.scala:
```diff
@@ -83,43 +80,28 @@ object CassandraEventsSink {
 
-      eventsRepartitioned.joinWithCassandraTable(
+      // [new events] = [batch] - [events table]
+      eventsRepartitioned.leftJoinWithCassandraTable(
```
Could we add a dependency timer for this so that we know how expensive it is to perform the join? This looks like a likely bottleneck that may want to be addressed in the future (e.g. using a different store for the already ingested event ids like Redis).

kevinhartman replied:
Unfortunately we can't time this since it's a Spark transformation (lazy). The Spark UI should have timing information (hopefully granular enough for us to see if this is a problem). It should be possible to push the `where` clause down to Cassandra itself as well, but I wasn't quickly able to figure out how to achieve that with the Cassandra connector. Redis would likely be much faster even still.

c-w committed:
Merge branch 'master' into counters

c-w committed:
Fix bad merge

c-w committed:
Refactor conjunctive topic tests to use RDDs

c-w committed:
Remove no longer useful test

The functionality should now be covered by the
CassandraConjunctiveTopics tests.

c-w committed:
Revert debug replication factor

c-w committed:
Merge branch 'master' into counters

c-w committed:
Fix CQL comment syntax highlighting in VSCode

c-w committed:
Remove comment addressed in pull request review

See https://github.com/CatalystCode/project-fortis/pull/31/files#r170260887

c-w committed:
Make comments consistent

c-w commented:
This is all looking pretty good.

![image](https://user-images.githubusercontent.com/1086421/36678117-e4c8c8f2-1add-11e8-8ffc-d5482dd6aa3a.png)

Going to merge this for now, please open follow-up PRs as you get to the items that are still outstanding.
